File: 프로그래머스/3/181186. 아방가르드 타일링/아방가르드 타일링.py
def solution(n):
    divide_num = 1000000007
    unique_num = [12, 2, 4]

    dp = [1, 1, 3, 10]

    for i in range(4, n + 1):
        basic_num = dp[i - 1] + 2 * dp[i - 2] + 5 * dp[i - 3]
        
        if i > 6:
            unique_num[i % 3] += 2 * dp[i - 4] + 2 * dp[i - 5] + 4 * dp[i - 6]
            unique_num[i % 3] %= divide_num

        dp.append((basic_num + unique_num[i % 3]) % divide_num)

    return dp[n]

File: 프로그래머스/3/181186. 아방가르드 타일링/아방가르드 타일링.java
import java.util.*;

class Solution {
    public int solution(int n) {
        int answer = 0;
        long divideNum = 1000000007;

        List<Integer> dp = new ArrayList<>();
        dp.add(1);
        dp.add(1);
        dp.add(3);
        dp.add(10);
        dp.add(23);
        dp.add(62);


        for (int i=6; i<n+1; i++) {
            long basicNum = (dp.get(i-1) + 2L *dp.get(i-2) + 6L *dp.get(i-3) + dp.get(i-4) - dp.get(i-6) + divideNum)%divideNum;

            dp.add((int) basicNum);
        }

        answer = dp.get(n);

        return answer;
    }
}